Add typing to deCONZ Lock and Logbook platforms

Kane610 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Start improving type hints to deCONZ integration. Does not depend on library for type hint, that's something for later.

Split from https://github.com/home-assistant/core/pull/58968

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Kane610 committed:
Add typing to deCONZ Lock and Logbook platforms

allenporter commented on homeassistant/components/deconz/lock.py:
```diff
@@ -72,16 +86,18 @@ class DeconzLock(DeconzDevice, LockEntity):
-    def is_locked(self):
+    def is_locked(self) -> bool:
         """Return true if lock is on."""
+        assert isinstance(self._device.is_locked, bool)
```
Given that _device is type and is_locked is typed to return a bool, why is this assert needed? It seems like there should be enough information to infer this type without that, right?

Kane610 replied:
This is due to the library not being used to trace types, I haven't added py.typed to it yet since that would require a lot of more rewrites, so Im taking it in steps.

frenck replied:
Fix the library instead. `ininstance` checks are runtime. This now adds runtime load for typing. Which IMHO is not a good solution.

Kane610 replied:
Based on feedback in other PRs I'm working through everything and removing all asserts

Kane610 replied:
Fixed

allenporter commented on homeassistant/components/deconz/logbook.py:
```diff
@@ -131,9 +131,10 @@ def async_describe_events(
+        deconz_alarm_event: DeconzAlarmEvent | DeconzEvent | None = (
+            _get_deconz_event_from_device_id(hass, event.data[ATTR_DEVICE_ID])
         )
+        assert deconz_alarm_event
```
Given the other comments in the other typing PRs about avoiding asserts: Do you have thoughts about another way to handle this? One thought is to always return a non-None object from `_get_deconz_event_from_device_id` and raise `InvalidDeviceAutomationConfig` when not found, like in some of the other cases.

Kane610 replied:
You're further ahead than me :), this is a good suggestion!

Kane610 replied:
I redid it according to your suggestion

frenck commented on homeassistant/components/deconz/logbook.py:
```diff
@@ -148,6 +149,7 @@ def async_describe_deconz_event(event: Event) -> dict:
         deconz_event: DeconzEvent | None = _get_deconz_event_from_device_id(
             hass, event.data[ATTR_DEVICE_ID]
         )
+        assert deconz_event
```
Why is this assert needed? Avoid the use of asserts.

Kane610 replied:
Fixed

Kane610 committed:
Improve typing of lock

Kane610 committed:
Improve hints
Remove asserts
Change _get_deconz_event_from_device_id to always return a deconz event

Kane610 committed:
Remove assert

allenporter commented:
Also appears to be a couple test failures

frenck reviewed (approved):
Thanks, @Kane610 👍

Kane610 commented:
Woohoo! Thanks @frenck 🎉
